File: server/src/_shared/packet.py
"""Signed packet envelope format used between buds-sim → host → server.

For Sprint 1 we use a deliberately simple JSON envelope (not yet the
full TLV from packet.html). This keeps debugging trivial while we wire
up the auth + transport path. We'll switch the body to TLV+Opus in
Sprint 2 without changing the signature scheme.

Wire format — three independent audio tracks:

    {
      "v":        1,
      "device":   "buds-demo-001",
      "session":  "01HZAB...",
      "ts":       1716123456789,
      "trigger":  "long_press",
      "tracks": {
        # what the buds were PLAYING into the user's ear
        "system":   { "codec": "pcm16", "duration_ms": 30000, "audio_b64": "...", "sha256": "..." },

        # what the OUTSIDE world sounded like (ANC mic on real buds,
        # system mic in the simulator)
        "external": { "codec": "pcm16", "duration_ms": 30000, "audio_b64": "...", "sha256": "..." },

        # what the USER said after the trigger fired (VAD-bounded)
        "question": { "codec": "pcm16", "duration_ms":  2300, "audio_b64": "...", "sha256": "..." }
      },
      "sig":      "<base64url(ed25519(canonical_bytes(body_without_sig)))>"
    }

The three tracks are independent so the server can choose which to
transcribe / fingerprint / ignore per intent. "방금 그 노래 뭐였어?"
analyzes the system track; "옆에서 누가 뭐라 했지?" analyzes the
external track; both also consider the question track for context.

`canonical_bytes()` produces a deterministic byte string the same way
on Python and TypeScript: JSON with sorted keys, no spaces, UTF-8.
"""
from __future__ import annotations

import json
from typing import Any, TypedDict


class Track(TypedDict):
    codec: str
    duration_ms: int
    audio_b64: str
    sha256: str


class Envelope(TypedDict, total=False):
    v: int
    device: str
    session: str
    ts: int
    trigger: str
    tracks: dict[str, Track]
    sig: str  # base64url(signature) — added by sign_envelope


def canonical_bytes(envelope: dict[str, Any]) -> bytes:
    """Deterministic JSON serialization for signing.

    - keys sorted
    - separators with no spaces  (',':' → ',',':')
    - ASCII-safe (ensure_ascii=False allows UTF-8; but no payload has
      non-ASCII identifiers, so this is just future-proofing)
    - `sig` field excluded (you sign the body, not the signature itself)
    """
    body = {k: v for k, v in envelope.items() if k != "sig"}
    return json.dumps(
        body,
        sort_keys=True,
        ensure_ascii=False,
        separators=(",", ":"),
    ).encode("utf-8")


def sign_envelope(envelope: dict[str, Any], keypair) -> dict[str, Any]:
    """Returns a new dict with `sig` appended."""
    msg = canonical_bytes(envelope)
    signed = dict(envelope)
    signed["sig"] = keypair.sign_b64(msg)
    return signed


def verify_envelope(envelope: dict[str, Any], public_b64: str) -> bool:
    """True iff `envelope['sig']` is a valid Ed25519 signature."""
    from .key import verify
    sig = envelope.get("sig")
    if not isinstance(sig, str) or not sig:
        return False
    msg = canonical_bytes(envelope)
    return verify(public_b64, msg, sig)

File: shared/ts/packet.js
// Signed packet envelope — TS/JS side.
//
// Mirrors shared/python/packet.py exactly so the server (Python)
// verifies signatures produced here (browser) without ambiguity.
//
// Canonical-bytes contract:
//   1. Take the envelope, drop the `sig` field.
//   2. JSON.stringify with keys deeply sorted, separators (",",":").
//   3. UTF-8 encode the resulting string.
//   4. That byte string is what gets Ed25519-signed.

export function canonicalBytes(envelope) {
  const withoutSig = {};
  for (const k of Object.keys(envelope)) {
    if (k !== "sig") withoutSig[k] = envelope[k];
  }
  const json = stableStringify(withoutSig);
  return new TextEncoder().encode(json);
}

function stableStringify(v) {
  if (v === null || typeof v !== "object") return JSON.stringify(v);
  if (Array.isArray(v)) {
    return "[" + v.map(stableStringify).join(",") + "]";
  }
  const keys = Object.keys(v).sort();
  return "{" + keys.map(k =>
    JSON.stringify(k) + ":" + stableStringify(v[k])
  ).join(",") + "}";
}

export function signEnvelope(envelope, keypair) {
  const msg = canonicalBytes(envelope);
  return { ...envelope, sig: keypair.signB64(msg) };
}

// Helper for buds-sim — produces an envelope from raw track data
// without yet doing Opus encoding (Sprint 2 swaps in real codec).
export function buildEnvelope({
  deviceId,
  sessionId,
  trigger = "long_press",
  lookbackBytes,    // Uint8Array — raw PCM placeholder
  questionBytes,    // Uint8Array
  lookbackMs,
  questionMs,
}) {
  return {
    v: 1,
    device: deviceId,
    session: sessionId,
    ts: Date.now(),
    trigger,
    tracks: {
      lookback: {
        codec: "pcm16",
        duration_ms: lookbackMs,
        audio_b64: bytesToB64u(lookbackBytes),
        sha256: "placeholder",
      },
      question: {
        codec: "pcm16",
        duration_ms: questionMs,
        audio_b64: bytesToB64u(questionBytes),
        sha256: "placeholder",
      },
    },
  };
}

function bytesToB64u(bytes) {
  if (!bytes || bytes.length === 0) return "";
  // Avoid stack blow-up on very large arrays
  let bin = "";
  const CHUNK = 0x8000;
  for (let i = 0; i < bytes.length; i += CHUNK) {
    bin += String.fromCharCode.apply(null, bytes.subarray(i, i + CHUNK));
  }
  return btoa(bin).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}
